File: website/docs-examples/3d/c/examples/vehicle_controller3.c
#include "snippets.h"

int main(void) {
    snippets_init(); /* aborts on any Rapier error, so the ctest fails */
    R3World *world = r3NewWorld();
    R3ColliderDesc ground = r3CuboidColliderDesc(r3Vector(100.0, 0.1, 100.0));
    r3InsertColliderWithoutParent(world, &ground);

    // DOCUSAURUS: Vehicle start
    // The chassis is an ordinary dynamic rigid-body.
    const R3Real hw = 0.3;
    const R3Real hh = 0.15;
    R3RigidBodyDesc chassis_body = r3DynamicRigidBodyDesc();
    chassis_body.position.translation = r3Vector(0.0, 1.0, 0.0);
    R3RigidBodyHandle chassis_handle = r3InsertRigidBody(world, &chassis_body);
    R3ColliderDesc chassis_collider = r3CuboidColliderDesc(r3Vector(hw * 2.0, hh, hw));
    chassis_collider.density = 100.0;
    r3InsertCollider(chassis_handle, &chassis_collider);

    // The tuning shared by the wheels: the suspension and the grip.
    R3WheelTuning tuning = r3DefaultWheelTuning();
    tuning.suspension_stiffness = 100.0;
    tuning.suspension_damping = 10.0;

    // The controller must be freed (with r3FreeDynamicRayCastVehicleController) before its world.
    R3DynamicRayCastVehicleController *vehicle = r3NewDynamicRayCastVehicleController(chassis_handle);
    const R3Vector wheel_positions[4] = {
        {hw * 1.5, -hh, hw},
        {hw * 1.5, -hh, -hw},
        {-hw * 1.5, -hh, hw},
        {-hw * 1.5, -hh, -hw},
    };

    for (size_t i = 0; i < 4; i++) {
        // The position of the wheel, the direction its suspension pushes along, its axle, the
        // rest length of its suspension, and its radius; all in the local frame of the chassis.
        r3DynamicRayCastVehicleController_AddWheel(vehicle, wheel_positions[i], r3Vector(0.0, -1.0, 0.0),
                                                   r3Vector(0.0, 0.0, 1.0), hh, hh / 4.0, &tuning);
    }
    // DOCUSAURUS: Vehicle stop

    // DOCUSAURUS: VehicleUpdate start
    // The wheels are ray-casted against the scene: the chassis itself is always excluded, and
    // every other dynamic body is generally excluded from these ray-casts too.
    R3QueryOptions options = r3DefaultQueryOptions();
    options.filter.flags = R3_QUERY_EXCLUDE_DYNAMIC;

    for (int i = 0; i < 200; i++) {
        // The vehicle is driven by setting the steering angle, the engine force, and the brake of
        // its wheels. Here the two front wheels (indices 0 and 1) are the driving and steering ones.
        r3DynamicRayCastVehicleController_SetWheelControls(vehicle, 0, 0.2, 30.0, 0.0);
        r3DynamicRayCastVehicleController_SetWheelControls(vehicle, 1, 0.2, 30.0, 0.0);

        r3DynamicRayCastVehicleController_UpdateVehicle(vehicle, r3TimeStep(world), &options);

        r3Step(world, NULL, NULL);
    }

    printf("Vehicle speed: %f\n", (double)r3DynamicRayCastVehicleController_CurrentVehicleSpeed(vehicle));
    // DOCUSAURUS: VehicleUpdate stop

    // DOCUSAURUS: VehicleWheels start
    // The wheels are given in the order they were added to the controller.
    R3WheelState wheels[4];
    size_t num_wheels = r3DynamicRayCastVehicleController_Wheels(vehicle, wheels, 4);
    for (size_t i = 0; i < num_wheels; i++) {
        // The world-space center of the wheel, its current suspension length, rotation angle, etc.
        printf("Wheel %zu: center (%f, %f, %f), suspension length %f, rotation %f, in contact: %u\n", i,
               (double)wheels[i].center.x, (double)wheels[i].center.y, (double)wheels[i].center.z,
               (double)wheels[i].suspension_length, (double)wheels[i].rotation,
               (unsigned)wheels[i].is_in_contact);
    }
    // DOCUSAURUS: VehicleWheels stop

    r3FreeDynamicRayCastVehicleController(vehicle);
    r3FreeWorld(world);
    return EXIT_SUCCESS;
}

File: website/docs-examples/3d/bevy/examples/vehicle_controller3.rs
use bevy::prelude::*;
use bevy_rapier3d::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_plugins(RapierPhysicsPlugin::<NoUserData>::default())
        .add_plugins(RapierDebugRenderPlugin::default())
        .add_systems(Startup, setup_graphics)
        .add_systems(Startup, setup_physics)
        .add_systems(FixedUpdate, drive_vehicle)
        .run();
}

fn setup_graphics(mut commands: Commands) {
    // Add a camera so we can see the debug-render.
    commands.spawn((
        Camera3d::default(),
        Transform::from_xyz(-3.0, 3.0, 10.0).looking_at(Vec3::ZERO, Vec3::Y),
    ));
}

fn setup_physics(mut commands: Commands) {
    /* Create the ground. */
    commands
        .spawn(Collider::cuboid(100.0, 0.1, 100.0))
        .insert(Transform::from_xyz(0.0, -0.1, 0.0));

    // DOCUSAURUS: Vehicle start
    // The tuning shared by the wheels: the suspension and the grip.
    let tuning = WheelTuning {
        suspension_stiffness: 100.0,
        suspension_damping: 10.0,
        ..WheelTuning::default()
    };

    let hw = 0.3;
    let hh = 0.15;
    let wheel_positions = [
        Vec3::new(hw * 1.5, -hh, hw),
        Vec3::new(hw * 1.5, -hh, -hw),
        Vec3::new(-hw * 1.5, -hh, hw),
        Vec3::new(-hw * 1.5, -hh, -hw),
    ];
    let wheels = wheel_positions
        .into_iter()
        .map(|position| {
            // The position of the wheel, the direction its suspension pushes along, its axle, the
            // rest length of its suspension, and its radius; all in the local frame of the chassis.
            VehicleWheel::new(position, -Vec3::Y, Vec3::Z, hh, hh / 4.0, tuning)
        })
        .collect();

    // The chassis is an ordinary dynamic rigid-body, with the vehicle controller attached to it.
    commands.spawn((
        Transform::from_xyz(0.0, 1.0, 0.0),
        RigidBody::Dynamic,
        Collider::cuboid(hw * 2.0, hh, hw),
        ColliderMassProperties::Density(100.0),
        RayCastVehicleController::new(wheels),
    ));
    // DOCUSAURUS: Vehicle stop
}

// DOCUSAURUS: VehicleUpdate start
fn drive_vehicle(mut vehicles: Query<&mut RayCastVehicleController>) {
    for mut vehicle in vehicles.iter_mut() {
        // The vehicle is driven by setting the engine force, the brake, and the steering angle of
        // its wheels. Here the two front wheels are the driving and steering ones.
        for wheel in &mut vehicle.wheels[0..2] {
            wheel.engine_force = 30.0;
            wheel.steering = 0.2;
        }

        // The colliders of the chassis are always ignored by the ray-casts of the wheels. Other
        // colliders, here the ones attached to dynamic rigid-bodies, can be ignored too.
        vehicle.filter_flags =
            QueryFilterFlags::EXCLUDE_SENSORS | QueryFilterFlags::EXCLUDE_DYNAMIC;

        // The results of the last update are written back into the component.
        println!("Vehicle speed: {}", vehicle.current_vehicle_speed);
        for wheel in &vehicle.wheels {
            println!(
                "Wheel in contact: {}, suspension length: {}",
                wheel.state.is_in_contact, wheel.state.suspension_length
            );
        }
    }
}
// DOCUSAURUS: VehicleUpdate stop
